python/ray/dashboard/metrics_exporter/client.py

```diff
@@ -0,0 +1,83 @@
+import logging
+import traceback
+
+from ray.dashboard.metrics_exporter import api
+from ray.dashboard.metrics_exporter.exporter import Exporter
+
+logger = logging.getLogger(__name__)
+
+
+class MetricsExportClient:
+    """Group of functionalities used by Dashboard to do external communication.
+
+    start_export_metrics should not be called more than once as it can create
+    multiple threads that export the same metrics.
+
+    Args:
+        address: Address to export metrics
+        dashboard_controller(BaseDashboardController): Dashboard controller to
+            run dashboard business logic.
+        dashboard_id(str): Unique dashboard ID.
+    """
+
+    def __init__(self, address, dashboard_controller, dashboard_id,
+                 exporter: Exporter):
+        self.dashboard_id = dashboard_id
+        self.auth_url = "{}/auth".format(address)
+        self.dashboard_controller = dashboard_controller
+        self.exporter = exporter
+
+        # Data obtained from requests.
+        self._dashboard_url = None
+        self.auth_info = None
+
+        # Client states
+        self.is_authenticated = False
+        self.is_exporting_started = False
+
+    def _authenticate(self):
+        """
+        Return:
+            Whether or not the authentication succeed.
+        """
+        self.auth_info = api.authentication_request(self.auth_url,
+                                                    self.dashboard_id)
+        self._dashboard_url = self.auth_info.dashboard_url
+        self.is_authenticated = True
+
+    @property
+    def enabled(self):
+        return self.is_authenticated
+
+    @property
+    def dashboard_url(self):
+        # This function should be used only after authentication succeed.
+        assert self._dashboard_url is not None, (
+            "dashboard url should be obtained by "
+            "`start_exporting_metrics` method first.")
+        return self._dashboard_url
+
+    def start_exporting_metrics(self):
+        """Create a thread to export metrics.
+
+        Once this function succeeds, it should not be called again.
+
+        Return:
+            Whether or not it suceedes to run exporter.
+        """
+        assert not self.is_exporting_started
+        if not self.is_authenticated:
+            try:
+                self._authenticate()
+            except Exception as e:
```

Comment:
In general, we should try to avoid catching Exceptions generically like this and catch more specific ones, as it can mask errors.

Reply:
TL;DR I wanted to just provide an instruction to try reenabling instead of breaking the whole local dashboard when authentication fails (even for unexpected exceptions). If you think it is a bad idea, I will revert it to only catch network related exceptions. (requests.exceptions.HTTPError and requests.exceptions.ConnectionError) 

Reasoning: 
This was indeed intended to prevent errors from happening. I don't want the local dashboard to be broken when client authentication fails (because if it happens, users will just turn off flags and never use this feature again).